### ISSUE TITLE
CodeMirror is not defined ckeditor 4.4.7

### ISSUE THREAD

**Issue by alexander-schranz** [w8tcha/CKEditor-CodeMirror-Plugin#77]
I get this error when I click on the sourcecode button

![CodeMirror is not defined](https://cloud.githubusercontent.com/assets/1698337/6331101/8e79c496-bb7d-11e4-9512-a84f6ed5dda5.png)

on the click on the button I see the scripts are loaded in the network but maybe to late?

![bildschirmfoto 2015-02-23 um 17 03 04](https://cloud.githubusercontent.com/assets/1698337/6331157/f0a8acb8-bb7d-11e4-9737-3a99fc279b69.png)

**Comment by SplicePHP** [comment]
I am having the exact same issue.
The problem only occurs when I try to load CKEditor via requirejs.
This never used to be an issue and it only started happening now.
Not sure where the problem lies.
Any help would be greatly appreciated.

**Comment by simshaun** [comment]
Same problem here.

I'm fairly sure it's occurring because the CodeMirror library detects the AMD loader and returns itself as an AMD module (like many other JS libs do). The CKEditor plugin expects CodeMirror to be a global, not an AMD module.

I'll attempt to fix this in a bit and submit a PR.

**Comment by StrawBerryAtWork** [comment]
Hi !
I got the same problem with Dojo 1.10.4 and ckEditor.
Any news about this problem ?

**Comment by w8tcha** [comment]
No sorry, i need some help with that one, because i have no experience with requirejs

**Comment by andrewboni** [comment]
Yes, running into this same issue with requirejs and this plugin

**Comment by andrewboni** [comment]
@SplicePHP were you able to find a workaround?

**Comment by istrasoft** [comment]
hi @w8tcha, thanks for the fix ! how/when will this be included in the CKEditor build that can be downloaded from their website ? or how can we already include this fix in an existing CKEditor implementation please ? 
Thanks

**Comment by w8tcha** [comment]
I published version 1.17 to the ckeditor.com site. You can now include the plugin with the Builder

**Comment by istrasoft** [comment]
Hello @w8tcha, just updated the whole build from ckeditor website and still have 

```
ckeditor.js:618 Uncaught ReferenceError: CodeMirror is not defined
    at l (ckeditor.js:618)
    at Object.<anonymous> (ckeditor.js:642)
    at f (ckeditor.js:247)
    at Array.D (ckeditor.js:247)
    at w (ckeditor.js:247)
    at ckeditor.js:248
```

as soon as I click the "Source button" ... ckeditor is loaded with a shim
```
	shim: {
		'ckeditor': {
			exports: 'CKEDITOR'
		},
```

**Comment by w8tcha** [comment]
@fanick1 can you help?

**Comment by istrasoft** [comment]
Is there maybe something I could add to require config to fix that in another way, I don't know 
... ?

**Comment by fanick1** [comment]
@istrasoft I'd recommend first checking that all the javascripts successfully load
e.g.:

- ...ckeditor/plugins/codemirror/js/codemirror.min.js
- ...ckeditor/plugins/codemirror/js/codemirror.addons.min.js
- ...ckeditor/plugins/codemirror/js/codemirror.mode.htmlmixed.min.js (for example)
- ...ckeditor/plugins/codemirror/js/codemirror.addons.search.min.js (for example)

(Could you perhaps specify which javascripts get downloaded in your case? It is specific case by case depending on the editing mode you use.)

I also recommend to double check that you really updated the ckeditor. I've just downloaded fresh version of ckeditor from the ckeditor builder website (including the codemirror plugin v1.17 ) and it works on my machine.

For further troubleshooting, could you copy-paste the builder url from the build-config.js located in the downloaded zip file (for me to get the same package as you have)?

**Comment by istrasoft** [comment]
@fanick1 thanks for your help !
So, here is what is loaded. 

![image](https://user-images.githubusercontent.com/1440900/29970606-899e36b8-8f25-11e7-92cb-09088df9e9f1.png)

To be noted that only everything up until timestamp.png is loaded initally, the rest is loaded only after I click on the "Source" button.

It is indeed a fresh updated CKEditor, in ckeditor.js : timestamp:"H7HD",version:"4.7.2",revision:"c9b79c9"
...
CKEDITOR.plugins.add("codemirror",{version:1.17,init:function(b)
...

The builder URL is : http://ckeditor.com/builder/496025135952e94e02ed0b6ff1c2f929

Thanks !

**Comment by fanick1** [comment]
The codemirror is supposed to load said sources after clicking on the "Source" button, that is correct.
It looks like in your case the files do not get downloaded via the require.js at all - from looking at the "Initiator" column in the network panel it gets loaded via the CKEditor through its custom loader thing. Normally it would display require.js for the discussed files.
The codemirror plugin is checking for the presence of the require function in which case the require.js should be used to download the dependencies - the check for the require must be turning out negative here. 
Perhaps verify that `"function" == typeof require` in the console, maybe you have require.js set up in some way that the plugin does not account for in its check.

**Comment by istrasoft** [comment]
Ah okay, thanks for the tip. Indeed I have a slighly modified version to avoid confusion for our developers, so the local require is "require" and the global one is only "requirejs" (while the standard requirejs version exposes both "require" and "requirejs" globally). Interestingly enough it's like that for 3 years in our product and it's working fine with previous versions of CKEditor and CodeMirror plugin, seems like there has been a change in the way the loader is used or something else. I'll check monday if changing the test to use the global "requirejs" in our case will solve the issue. Thanks a lot already @fanick1

**Comment by istrasoft** [comment]
hi @fanick1 & @w8tcha. Solved my issue thanks to your input, thanks. There were two parts actually.
First was indeed to replace the 3-4 instances of calls to "require" by "requirejs" which was specific to my version.
However after that, CKEDITOR_GETURL was not defined somehow, and indeed this function wasnt declared in my build of ckeditor. So, I changed that single call to this and now it's fine. juts FYI ;)
`var g=window.CKEDITOR.getUrl("plugins/codemirror/js").replace(/\?.*$/,"");`

Thanks again for your support !
